### DISCUSSION TITLE
How to best use plotter.clear()?

### DISCUSSION THREAD

**Discussion by gheylam** [marcomusy/vedo#643] [Q&A] [answered]
I have been using vedo for animating moving graphs for some time now. However, I finding that different computers causes Plotter.show() to behave differently. 

For one computer, when I use Plotter.show(), the entire renderer is rerendered. So if I originally have 3 points, then show 2 points, the original 3 points will disappear. 
For a second computer, when I use Plotter.show(), the existing actors will remain and I will see 5 points. 

Therefore to make the behaviour more predictable, I am planning to use Plotter.clear() to remove all the actors. However, after using clear, all my buttons and axes are gone and I cannot add them back in. I.e here is an example:

```
 from vedo import * 
 import numpy as np
 
 # Create points 
 x_pt = Point([1, 0, 0], c='o5', r=8)
 y_pt = Point([0, 1, 0], c='g5', r=8)
 z_pt = Point([0, 0, 1], c='r5', r=8)

 # Show 3 points (at this stage, axes will be shown) 
 vedo_plt.show(x_pt, y_pt, z_pt, f"3 Points", axes=1, interactive=False)

 # Perform the clear
 vedo_plt.clear()

 # Show one point with the same axes 
 vedo_plt.show(x_pt, f"1 Point", axes=1, interactive=False)
 # Force the axes the appear
 vedo_plt.addGlobalAxes(axtype=1)
 vedo_plt.interactive()
```

I want to reuse the same plotter and be able to clear it content and show new content. But the trouble is that my buttons and axes simply can't return to the same plotter. I was hoping someone could tell me what they're doing to create animations in vedo or whether i'm using Plotter.clear() incorrectly. Would appreciate any insights, thank you!

**Comment by marcomusy** [comment, marked as answer]
Thanks for reporting this, can you tell me on which operating system you observe these different behaviours?
What vedo/vtk versions are you using? (type `vedo --info`)

You made a good point that the `clear()` method should not get rid of the buttons! I'll try to see if I can fix that.
In any case it should not be necessary to use clear. In this example you can create the `Axes` manually and treat them like any other vedo object in the scene.

```python
from vedo import *

vedo_plt = Plotter(interactive=False)

# Create points
x_pt = Point([1, 0, 0], c='o5', r=20)
y_pt = Point([0, 1, 0], c='g5', r=20)
z_pt = Point([0, 0, 1], c='r5', r=20)

# operation + creates an Assembly object, but you can also
# explicitly give the bounds you wish with xrange=... etc
axes = Axes( x_pt + y_pt + z_pt )

# Show 3 points (at this stage, axes will be shown)
vedo_plt.show(x_pt, y_pt, z_pt, axes, "3 Points").interactive()

# Perform the clear
vedo_plt.clear()

# Show one point with the same axes
vedo_plt.show(x_pt, axes, "1 Point")

vedo_plt.interactive()
```

In general, once the scene is rendered with `show()`, if you have loops or callbacks to update your scene, the best practice is to just  use `plt.add()` and `plt.remove()`  as needed (note that `add()` automatically makes a call to `.render()`).
Remember that you can give a name to any objects eg.: 
`my_sphere.name = "MySphere1"` and then remove it in a callback function with `.remove("MySphere1")`

**Reply by gheylam** [reply]
Hey marcomusy, thank you for your response and the recommendation regarding the plt.add() and plt.remove. Also your method of using the explicit axes creation works great. 

As for the different operating systems (or machines) which I can see different behaviours of plt.show(): 
- (Denote machine A) For the plt.show() which removes all existing actors for new ones: 
  vedo version      : 2022.2.0   https://vedo.embl.es
  vtk version       : 9.0.3
  python version    : 3.8.13 (default, May  9 2022, 16:01:31) [MSC v.1929 64 bit (AMD64)]
  system            : Windows 10 nt AMD64

- (Denote machine B) For the plt.show() which does not remove the existing actors for new ones: 
  vedo version      : 2022.2.3   https://vedo.embl.es
  vtk version       : 9.0.3
  python version    : 3.8.13 (default, May  9 2022, 16:01:31) [MSC v.1929 64 bit (AMD64)]
  system            : Windows 10 nt AMD64

From this, I install vedo version 2022.2.3 on my machine B and plt.show() now works as expected. Thank you for resolving this issue on my end!

**Reply by marcomusy** [reply]
Thanks you @gheylam for the detailed report.. indeed the fault is on my side because I didn't document the change of behavior of `show()` in the last release.
I´m glad that it works for you.
